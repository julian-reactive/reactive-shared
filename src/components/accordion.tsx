// Libraries
import React from 'react'
import map from 'lodash/map'

// Material Components
import Accordion, { AccordionProps } from '@mui/material/Accordion'
import AccordionSummary from '@mui/material/AccordionSummary'
import AccordionDetails from '@mui/material/AccordionDetails'
import Paper from '@mui/material/Paper'
import Box from '@mui/material/Box'

// Icons
import ExpandMoreIcon from '@mui/icons-material/ExpandMore'

interface ItemsProps extends AccordionProps {
  /**
     * Array of {name, children}
     *
     * @param {string} name: title of accordion tab
     * @param {ReactElement} children: inherited component. will be inside of accordion tab
     */
  name: string
}

export type SharedAccordionProps = { items?: ItemsProps[] } | null

const SharedAccordionComponent: React.FC<SharedAccordionProps> = props => {
  return (
    <Paper>
      {map(props?.items, ({ name, children, ...props }, itemIdx) => (
        <Accordion
          key={`common-accordion-item-${itemIdx}-${name}-${props.id}`}
          {...props}
        >
          <AccordionSummary
            aria-controls={`panel${itemIdx}a-content`}
            expandIcon={<ExpandMoreIcon />}
            id={`panel${itemIdx}a-header`}
          >
            <Box sx={{ textTransform: 'capitalize', fontWeight: 'bold' }}>
              {name}
            </Box>
          </AccordionSummary>

          <AccordionDetails>
            <Box>
              {children}
            </Box>
          </AccordionDetails>
        </Accordion>
      ))}
    </Paper>
  )
}

export const SharedAccordion = React.memo(SharedAccordionComponent)
