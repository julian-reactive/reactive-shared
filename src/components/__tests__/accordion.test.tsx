import React from 'react'

import {render} from '@testing-library/react'
import { SharedAccordion, SharedAccordionProps } from '../accordion'

const defaultProps: SharedAccordionProps = {
    items: [
{
    name: 'test',
    children: <span>test</span>
}
    ]
    
}

const getRendered = (props = defaultProps) => render(<SharedAccordion {...props} />)

describe('accordion', () => {
    it('should match snapshot', () => {
        const { container } = getRendered()
        expect(container.firstChild).toMatchSnapshot()
    })
})