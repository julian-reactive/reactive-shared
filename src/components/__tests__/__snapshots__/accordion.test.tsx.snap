// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accordion should match snapshot 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1elwnq4-MuiPaper-root-MuiAccordion-root"
  >
    <div
      aria-controls="panel0a-content"
      aria-expanded="false"
      class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-sh22l5-MuiButtonBase-root-MuiAccordionSummary-root"
      id="panel0a-header"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
      >
        <div
          class="MuiBox-root css-10iahqc"
        >
          test
        </div>
      </div>
      <div
        class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
    </div>
    <div
      class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
        >
          <div
            aria-labelledby="panel0a-header"
            class="MuiAccordion-region"
            id="panel0a-content"
            role="region"
          >
            <div
              class="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
            >
              <div
                class="MuiBox-root css-0"
              >
                <span>
                  test
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
